satpy/modifiers/parallax.py

```diff
@@ -0,0 +1,503 @@
+    This is is the object returned by :meth:`corrected_area`.
+
+    This procedure can be configured as a modifier using the
+    :class:`ParallaxCorrectionModifier` class.  However, the modifier can only
+    be applied to one dataset at the time, which may not provide optimal
+    performance, although dask should reuse identical calculations between
+    multiple channels.
+
+    """
+
+    def __init__(self, base_area,
+                 debug_mode=False):
+        """Initialise parallax correction class.
+
+        Args:
+            base_area (:class:`~pyresample.AreaDefinition`): Area for which calculated
+                geolocation will be calculated.
+            debug_mode (bool): Store diagnostic information in
+                self.diagnostics.  This attribute always apply to the most
+                recently applied operation only.
+        """
+        self.base_area = base_area
+        self.debug_mode = debug_mode
+        self.diagnostics = {}
+
+    def __call__(self, cth_dataset):
+        """Apply parallax correction to dataset.
+
+        Args:
+            cth_dataset: Dataset containing cloud top heights (or other heights
+                to be corrected).
+
+        Returns:
+            :class:'~pyresample.geometry.SwathDefinition`: Swathdefinition with corrected
+                lat/lon coordinates.
+        """
+        self.diagnostics.clear()
+        return self.corrected_area(cth_dataset)
+
+    def corrected_area(self, cth_dataset):
+        """Return the parallax corrected SwathDefinition.
+
+        Using the cloud top heights provided in ``cth_dataset``, calculate the
+        :class:`pyresample.geometry.SwathDefinition` that estimates the
+        geolocation for each pixel if it had been viewed from straight above
+        (without parallax error).  The cloud top height will first be resampled
+        onto the area passed upon class initialisation in :meth:`__init__`.
+        Pixels that are invisible after parallax correction are not retained
+        but get geolocation NaN.
+
+        Args:
+            cth_dataset (:class:`~xarray.DataArray`): Cloud top height in
+                metres.  The variable attributes must contain an ``area``
+                attribute describing the geolocation in a pyresample-aware way,
+                and they must contain satellite orbital parameters.  The
+                dimensions must be ``(y, x)``.  For best performance, this
+                should be a dask-based :class:`~xarray.DataArray`.
+
+        Returns:
+            :class:`~pyresample.geometry.SwathDefinition` describing parallax
+            corrected geolocation.
+        """
+        logger.debug("Calculating parallax correction using heights from "
+                     f"{cth_dataset.attrs.get('name', cth_dataset.name)!s}, "
+                     f"with base area {self.base_area.name!s}.")
+        (sat_lon, sat_lat, sat_alt_m) = _get_satpos_from_cth(cth_dataset)
```

Comment:
Shouldn't we allow satpy to look for satpos info from the other datasets passed to the modifier? Of course, in some cases, only the cth will be passed, but in other cases the satellite position will be available in other datasets.

Reply:
Yes, we should, but I'm not sure about the API.  Ideally, the modifier would go through whatever other datasets are passed to it, but I don't think Satpy is quite flexible enough for that?  A modifier that can obtain the CTH from other datasets would have to look like this, and use the optional prerequisites exclusively for the metadata?

```
        modifiers:
          parallax_corrected:
            modifier: !!python/name:satpy.modifiers.parallax.ParallaxCorrectionModifier
            prerequisites:
              - "ctth_alti"
           optional_prerequisites:
              - "IR_108"
              - "IR_120"
```
We should also allow the user to pass it in satellite location hardcoded.

Reply:
I have added this as a future TODO, so I'm deferring this enhanced functionality to a PR that I can work on after this one is merged.